README.md
<h2 align="center">
  <br>
  <img src="resources/icon.png" height="60" width="60">
  <br>
  JioTV
  <br>
</h2>

<h4 align="center">JioTV Kodi Add-on</h4>

<br>

## Disclaimer

This plugin is not officially commissioned/supported by Jio. The trademark "Jio" is registered by "Reliance Corporate IT Park Limited (RCITPL)"

## Download

[**Download**](https://github.com/botallen/plugin.video.jiotv/releases/latest) the `.zip` file.


changelog.txt
[- 3.0.0 -]
[fixes] fixed broken play urls

[- 2.3.0 -]
[added] more extra channels
[added] inputstream adaptive as a dependency 

[- 2.2.0 -]
[fixed] Bugs related to playback
[added] Login with OTP

[- 2.1.0 -]

[added] Kodi Matrix support
[added] Star channels support

[- 2.0.14 -]

[added] Featured Content

[- 2.0.13 -]

[added] Web login

[- 2.0.0 -]

[fixed] All major issues related to playback almost solved
[added] Categorized by Language 
[added] Filter channels by specific language in genre. (from settings)

[- 1.0.8-2 -]

[fixed] jio server issue (minor)

[- 1.0.8 -]

[added] Star sports channels are working now #3 (major)
[added] Catchup support for upto 7 days of shows (major)
[added] categories icons (minor)

[- 1.0.7 -]

[added] Proxy service to improve quality of channels (major)

[- 1.0.6 -]

[added] Changed whole code base to kodi friendly and improved perfomace (major)
[fixed] Star channels were not working (minor)

[- 1.0.5 -]

[added] Improved performace by removing non-sense code (minor)
[fixed] PVR setup setting bugs (minor)

[- 1.0.4 -]

[fixed] PVR setup setting (minor)

[- 1.0.3 -]

Initial public

[- 1.0.2 -]

Initial public

[- 1.0.1 -]

Initial public

[- 1.0.0 -]

Initial development


service.py
# -*- coding: utf-8 -*-

from resources.lib import proxy
from codequick import Script
from codequick.script import Settings
from socketserver import ThreadingTCPServer
import threading
from xbmc import Monitor, executebuiltin
from kodi_six import xbmcgui


def serveForever(handler):
    try:
        handler.serve_forever()
    except Exception as e:
        Script.log(e, lvl=Script.DEBUG)
        pass


ThreadingTCPServer.allow_reuse_address = True
_PORT = 48996
handler = ThreadingTCPServer(("", _PORT), proxy.JioTVProxy)
t = threading.Thread(target=serveForever, args=(handler,))
t.setDaemon(True)
t.start()

if not Settings.get_boolean("popup"):
    xbmcgui.Dialog().ok("JioTV Notification",
                        "Now you can create your custom playlist from anderson Dashboard. [CR]Find out more at [B]https://anderson.com/#dashboard[/B] [CR][CR]If you like this add-on then consider donating from [B]https://anderson.com/#donate[/B] [CR][CR]Github: [B]https://github.com/anderson/repository.anderson[/B] [CR]Discord: [B]https://anderson.com/discord[/B] [CR][CR][I]You can disable this popup from settings[/I]")

if Settings.get_boolean("m3ugen"):
    executebuiltin(
        "RunPlugin(plugin://plugin.video.jiotv/resources/lib/main/m3ugen/?notify=no)")

monitor = Monitor()
while not monitor.abortRequested():
    if monitor.waitForAbort(10):
        handler.shutdown()
        handler.server_close()
        break


resources/lib/proxy.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

# basic imports
from http.server import SimpleHTTPRequestHandler
import os
from urllib.parse import parse_qs, urlparse
from xbmcvfs import translatePath
import xbmcaddon
from resources.lib.utils import login, sendOTP

# codequick imports
from codequick import Script

ADDON = xbmcaddon.Addon()
ADDON_PATH = ADDON.getAddonInfo("path")


class JioTVProxy(SimpleHTTPRequestHandler):

    def do_GET(self):
        path = urlparse(self.path).path
        if path == "/":
            self.send_response(200)

            html = os.path.join(translatePath(
                ADDON_PATH), "resources", "login.html")
            try:
                f = open(html, 'rb')
            except IOError:
                self.send_error(404, "File not found")
                return None

            self.send_header("Content-type", "text/html")
            fs = os.fstat(f.fileno())
            self.send_header("Content-Length", str(fs.st_size))
            self.end_headers()
            self.wfile.write(bytes(f.read()))
            f.close()
            return
        else:
            self.send_error(404, "File not found")

    def do_POST(self):
        if self.path == "/login":
            data_string = self.rfile.read(
                int(self.headers['Content-Length']))

            qs = parse_qs(data_string.decode('utf-8'))
            error = None
            Script.log(qs, lvl=Script.INFO)
            try:
                if qs.get("type")[0] == "password":
                    error = login(qs.get("username")[0], qs.get("password")[0])
                elif qs.get("type")[0] == "otp":
                    mobile = qs.get("mobile")[0]
                    if qs.get("otp"):
                        error = login(mobile, qs.get("otp")[0], mode="otp")
                    else:
                        error = sendOTP(mobile)
                else:
                    error = "Invalid Type"
            except Exception as e:
                Script.log(e, lvl=Script.ERROR)
                error = str(e)

            if error:
                location = "/?error="+str(error)
            elif qs.get("type")[0] == "otp" and qs.get("otp") is None:
                location = "/?otpsent=" + qs.get("mobile")[0]
            else:
                location = "/?success"
            self.send_response(302)
            self.send_header('Location', location)
            self.end_headers()
        else:
            self.send_error(404, "File not found")


resources/lib/utils.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

import warnings
import contextlib
import urlquick
import requests
from uuid import uuid4
import base64
import hashlib
import urllib3
import time
from urllib3.exceptions import InsecureRequestWarning
from functools import wraps
from distutils.version import LooseVersion
from codequick import Script
from codequick.storage import PersistentDict
from xbmc import executebuiltin
from xbmcgui import Dialog
import socket

urllib3.disable_warnings()

old_merge_environment_settings = requests.Session.merge_environment_settings

@contextlib.contextmanager
def no_ssl_verification():
    opened_adapters = set()

    def merge_environment_settings(self, url, proxies, stream, verify, cert):
        # Verification happens only once per connection so we need to close
        # all the opened adapters once we're done. Otherwise, the effects of
        # verify=False persist beyond the end of this context manager.
        opened_adapters.add(self.get_adapter(url))

        settings = old_merge_environment_settings(self, url, proxies, stream, verify, cert)
        settings['verify'] = False

        return settings

    requests.Session.merge_environment_settings = merge_environment_settings

    try:
        with warnings.catch_warnings():
            warnings.simplefilter('ignore', InsecureRequestWarning)
            yield
    finally:
        requests.Session.merge_environment_settings = old_merge_environment_settings

        for adapter in opened_adapters:
            try:
                adapter.close()
            except:
                pass

def get_local_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.settimeout(0)
    try:
        s.connect(("8.8.8.8", 80))
        IP = s.getsockname()[0]
    except Exception:
        IP = '127.0.0.1'
    finally:
        s.close()
    return IP


def isLoggedIn(func):
    """
    Decorator to ensure that a valid login is present when calling a method
    """
    @wraps(func)
    def login_wrapper(*args, **kwargs):
        # Script.log("####################################### CHECKING TOKEN #######################################", lvl=Script.ERROR)
        with PersistentDict("headers") as db:
            username = db.get("username")
            password = db.get("password")
            headers = db.get("headers")
            exp = db.get("exp", 0)
        if headers and exp > time.time():
            return func(*args, **kwargs)
        elif username and password:
            login(username, password)
            return func(*args, **kwargs)
        elif headers and exp < time.time():
            Script.notify(
                "Login Error", "Session expired. Please login again")
            executebuiltin(
                "RunPlugin(plugin://plugin.video.jiotv/resources/lib/main/login/)")
            return False
        else:
            Script.notify(
                "Login Error", "You need to login with Jio Username and password to use this add-on")
            executebuiltin(
                "RunPlugin(plugin://plugin.video.jiotv/resources/lib/main/login/)")
            return False
    return login_wrapper


def refresh_token():
    headers = getHeaders()
    if not headers:
        return
    url = "https://auth.media.jio.com/tokenservice/apis/v1/refreshtoken?langId=6"
    refresh_headers = {
        "accesstoken": headers.get("authToken",""),
        "uniqueid": headers.get("uniqueId",""),
        "devicetype":"phone",
        "os":"android",
        "user-agent":"okhttp/4.2.2",
        "versioncode": "285",
    }
    body = {
        "appName": "RJIL_JioTV",
        "deviceId": headers.get("deviceId",""),
        "refreshToken": headers.get("refreshToken","")
    }
    # Script.log("####################################### TOKEN REFRESH #######################################", lvl=Script.ERROR)
    # Script.log(url, lvl=Script.ERROR)
    # Script.log(refresh_headers, lvl=Script.ERROR)
    # Script.log(body, lvl=Script.ERROR)
    with no_ssl_verification():
        resp = urlquick.post(url, json=body, headers=refresh_headers, max_age=-1, verify=False, raise_for_status=False).json()
        if resp.get("authToken", "") != "":
            # Script.log("Token refreshed", lvl=Script.ERROR)            
            headers["authToken"] = resp.get("authToken")
            with PersistentDict("headers") as db:
                db["headers"] = headers
                db["exp"] = time.time() + 432000
        
    return None


def login(username, password, mode="unpw"):
    deviceId = "1e075302d2fb0b64"
    login_headers = {
        "appname":"RJIL_JioTV",
        "devicetype":"phone",
        "os":"android",
        "user-agent":"okhttp/4.2.2",
    }
    if "+91" not in username:
        username = "+91" + username
    body = {
        "number": base64.b64encode(username.encode()).decode(),
        "otp": password,
        "deviceInfo": {
            "consumptionDeviceName": "ONEPLUS A5000",
            "info": {
                "type": "android",
                "platform": {
                    "name": "OnePlus5",
                },
                "androidId": deviceId
            }
        }
    }
    # Script.log(body, lvl=Script.ERROR)
    resp = urlquick.post("https://jiotvapi.media.jio.com/userservice/apis/v1/loginotp/verify", json=body, headers=login_headers, max_age=-1, verify=False, raise_for_status=False).json()
    # Script.log(resp, lvl=Script.INFO)
    if resp.get("ssoToken", "") != "":
        _CREDS = {
            "authToken": resp.get("authToken"),
            "refreshToken": resp.get("refreshToken"),
            "jToken": resp.get("jToken"),
            "ssotoken": resp.get("ssoToken"),
            "userId": resp.get("sessionAttributes", {}).get("user", {}).get("uid"),
            "uniqueId": resp.get("sessionAttributes", {}).get("user", {}).get("unique"),
            "crmid": resp.get("sessionAttributes", {}).get("user", {}).get("subscriberId"),
        }
        headers = {
            "appkey": "NzNiMDhlYzQyNjJm",
            "deviceId": "1e075302d2fb0b64",            
            "User-Agent": "JioTV",
            "os": "Android",
            "deviceId": deviceId,
            "versionCode": "285",
            "devicetype": "phone",
            "srno": "200206173037",            
            "channelid": "100",
            "usergroup": "tvYR7NSNn7rymo3F",
            "lbcookie": "1"
        }
        headers.update(_CREDS)
        with PersistentDict("headers") as db:
            db["headers"] = headers
            db["exp"] = time.time() + 432000
            if mode == "unpw":
                db["username"] = username
                db["password"] = password
        Script.notify("Login Success", "")
        return None
    else:        
        msg = resp.get("message", "Unknow Error")
        Script.notify("Login Failed", msg)
        return msg


def sendOTP(mobile):
    url = "https://jiotvapi.media.jio.com/userservice/apis/v1/loginotp/send"
    login_headers = {
        "appname":"RJIL_JioTV",
        "devicetype":"phone",
        "os":"android",
        "user-agent":"okhttp/4.2.2",
        "Content-Type": "application/json",
    }
    if "+91" not in mobile:
        mobile = "+91" + mobile
    body = {"number": base64.b64encode(mobile.encode()).decode()}
    # Script.log(url, lvl=Script.ERROR)
    # Script.log(login_headers, lvl=Script.ERROR)
    # Script.log(body, lvl=Script.ERROR)
    resp = urlquick.post(url, json=body, headers=login_headers, verify=False, raise_for_status=False)    
    if resp.status_code == 400:
        Script.notify("Otp Send Failed", resp.get("message", ""))
        return resp.json().get("message")
    if resp.status_code == 204:
        Script.notify("Otp Sent", "")
        return None
    return None


def logout():
    with PersistentDict("headers") as db:
        del db["headers"]
    Script.notify("You\'ve been logged out", "")


def getHeaders():
    with PersistentDict("headers") as db:
        return db.get("headers", False)


def getTokenParams():
    def magic(x): return base64.b64encode(hashlib.md5(x.encode()).digest()).decode().replace(
        '=', '').replace('+', '-').replace('/', '_').replace('\r', '').replace('\n', '')
    pxe = str(int(time.time()+(3600*9.2)))
    jct = magic("cutibeau2ic9p-O_v1qIyd6E-rf8_gEOQ"+pxe)
    return {"jct": jct, "pxe": pxe, "st": "9p-O_v1qIyd6E-rf8_gEOQ"}


def check_addon(addonid, minVersion=False):
    """Checks if selected add-on is installed."""
    try:
        curVersion = Script.get_info("version", addonid)
        if minVersion and LooseVersion(curVersion) < LooseVersion(minVersion):
            # Script.log('{addon} {curVersion} doesn\'t setisfy required version {minVersion}.'.format(
                # addon=addonid, curVersion=curVersion, minVersion=minVersion))
            Dialog().ok("Error", "{minVersion} version of {addon} is required to play this content.".format(
                addon=addonid, minVersion=minVersion))
            return False
        return True
    except RuntimeError:
        Script.log('{addon} is not installed.'.format(addon=addonid))
        if not _install_addon(addonid):
            # inputstream is missing on system
            Dialog().ok("Error",
                        "[B]{addon}[/B] is missing on your Kodi install. This add-on is required to play this content.".format(addon=addonid))
            return False
        return True


def _install_addon(addonid):
    """Install addon."""
    try:
        # See if there's an installed repo that has it
        executebuiltin('InstallAddon({})'.format(addonid), wait=True)

        # Check if add-on exists!
        version = Script.get_info("version", addonid)

        Script.log(
            '{addon} {version} add-on installed from repo.'.format(addon=addonid, version=version))
        return True
    except RuntimeError:
        Script.log('{addon} add-on not installed.'.format(addon=addonid))
        return False


resources/lib/main.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

# xbmc imports
from xbmcaddon import Addon
from xbmc import executebuiltin
from xbmcgui import Dialog, DialogProgress

# codequick imports
from codequick import Route, run, Listitem, Resolver, Script
from codequick.utils import keyboard
from codequick.script import Settings
from codequick.storage import PersistentDict

# add-on imports
from resources.lib.utils import getTokenParams, getHeaders, isLoggedIn, refresh_token, login as ULogin, logout as ULogout, check_addon, sendOTP, get_local_ip, no_ssl_verification
from resources.lib.constants import GET_CHANNEL_URL, PLAY_EX_URL, EXTRA_CHANNELS, GENRE_MAP, LANG_MAP, FEATURED_SRC, CONFIG, CHANNELS_SRC, IMG_CATCHUP, PLAY_URL, IMG_CATCHUP_SHOWS, CATCHUP_SRC, M3U_SRC, EPG_SRC, M3U_CHANNEL

# additional imports
import urlquick
import requests
from urllib.parse import urlencode
import inputstreamhelper
import json
import m3u8
from time import time, sleep
from datetime import datetime, timedelta, date


# Root path of plugin
@Route.register
def root(plugin):
    yield Listitem.from_dict(**{
        "label": "Featured",
        "art": {
            "thumb": IMG_CATCHUP_SHOWS + "cms/TKSS_Carousal1.jpg",
            "icon": IMG_CATCHUP_SHOWS + "cms/TKSS_Carousal1.jpg",
            "fanart": IMG_CATCHUP_SHOWS + "cms/TKSS_Carousal1.jpg",
        },
        "callback": Route.ref("/resources/lib/main:show_featured")
    })
    for e in ["Genres", "Languages"]:
        yield Listitem.from_dict(**{
            "label": e,
            "art": {
                "thumb": CONFIG[e][0].get("tvImg"),
                "icon": CONFIG[e][0].get("tvImg"),
                "fanart": CONFIG[e][0].get("promoImg"),
            },
            "callback": Route.ref("/resources/lib/main:show_listby"),
            "params": {"by": e}
        })


# Shows Featured Content
@Route.register
def show_featured(plugin, id=None):
    resp = urlquick.get(FEATURED_SRC, headers={
        "usergroup": "tvYR7NSNn7rymo3F",
        "os": "android",
        "devicetype": "phone",
        "versionCode": "226"
    }, max_age=-1).json()
    for each in resp.get("featuredNewData", []):
        if id:
            if int(each.get("id", 0)) == int(id):
                data = each.get("data", [])
                for child in data:
                    info_dict = {
                        "art": {
                            "thumb": IMG_CATCHUP_SHOWS + child.get("episodePoster", ""),
                            "icon": IMG_CATCHUP_SHOWS + child.get("episodePoster", ""),
                            "fanart": IMG_CATCHUP_SHOWS + child.get("episodePoster", ""),
                            "clearart": IMG_CATCHUP + child.get("logoUrl", ""),
                            "clearlogo": IMG_CATCHUP + child.get("logoUrl", ""),
                        },
                        "info": {
                            'originaltitle': child.get("showname"),
                            "tvshowtitle": child.get("showname"),
                            "genre": child.get("showGenre"),
                            "plot": child.get("description"),
                            "episodeguide": child.get("episode_desc"),
                            "episode": 0 if child.get("episode_num") == -1 else child.get("episode_num"),
                            "cast": child.get("starCast", "").split(', '),
                            "director": child.get("director"),
                            "duration": child.get("duration")*60,
                            "tag": child.get("keywords"),
                            "mediatype": "movie" if child.get("channel_category_name") == "Movies" else "episode",
                        }
                    }
                    if child.get("showStatus") == "Now":
                        info_dict["label"] = info_dict["info"]["title"] = child.get(
                            "showname", "") + " [COLOR red] [ LIVE ] [/COLOR]"
                        info_dict["callback"] = play
                        info_dict["params"] = {
                            "channel_id": child.get("channel_id")}
                        yield Listitem.from_dict(**info_dict)
                    elif child.get("showStatus") == "future":
                        timetext = datetime.fromtimestamp(int(child.get("startEpoch", 0)*.001)).strftime(
                            '    [ %I:%M %p -') + datetime.fromtimestamp(int(child.get("endEpoch", 0)*.001)).strftime(' %I:%M %p ]   %a')
                        info_dict["label"] = info_dict["info"]["title"] = child.get(
                            "showname", "") + (" [COLOR green]%s[/COLOR]" % timetext)
                        info_dict["callback"] = ""
                        yield Listitem.from_dict(**info_dict)
                    elif child.get("showStatus") == "catchup":
                        timetext = datetime.fromtimestamp(int(child.get("startEpoch", 0)*.001)).strftime(
                            '    [ %I:%M %p -') + datetime.fromtimestamp(int(child.get("endEpoch", 0)*.001)).strftime(' %I:%M %p ]   %a')
                        info_dict["label"] = info_dict["info"]["title"] = child.get(
                            "showname", "") + (" [COLOR yellow]%s[/COLOR]" % timetext)
                        info_dict["callback"] = play
                        info_dict["params"] = {
                            "channel_id": child.get("channel_id"),
                            "showtime": child.get("showtime", "").replace(":", ""),
                            "srno": datetime.fromtimestamp(int(child.get("startEpoch", 0)*.001)).strftime('%Y%m%d')
                        }
                        yield Listitem.from_dict(**info_dict)
        else:
            yield Listitem.from_dict(**{
                "label": each.get("name"),
                "art": {
                    "thumb": IMG_CATCHUP_SHOWS + each.get("data", [{}])[0].get("episodePoster"),
                    "icon": IMG_CATCHUP_SHOWS + each.get("data", [{}])[0].get("episodePoster"),
                    "fanart": IMG_CATCHUP_SHOWS + each.get("data", [{}])[0].get("episodePoster"),
                },
                "callback": Route.ref("/resources/lib/main:show_featured"),
                "params": {"id": each.get("id")}
            })


# Shows Filter options
@Route.register
def show_listby(plugin, by):
    for each in CONFIG[by]:
        yield Listitem.from_dict(**{
            "label": each.get("name"),
            "art": {
                "thumb": each.get("tvImg"),
                "icon": each.get("tvImg"),
                "fanart": each.get("promoImg")
            },
            "callback": Route.ref("/resources/lib/main:show_category"),
            "params": {"category_id": each.get("name").replace(" ", ""), "by": by}
        })


# Shows channels by selected filter/category
@Route.register
def show_category(plugin, category_id, by):
    resp = urlquick.get(CHANNELS_SRC).json().get("result")

    def fltr(x):
        fby = by.lower()[:-1]
        if fby == "genre":
            return GENRE_MAP[x.get("channelCategoryId")] == category_id and Settings.get_boolean(LANG_MAP[x.get("channelLanguageId")])
        else:
            return LANG_MAP[x.get("channelLanguageId")] == category_id

    for each in filter(fltr, resp):
        if each.get("channelIdForRedirect") and not Settings.get_boolean("extra"):
            continue
        litm = Listitem.from_dict(**{
            "label": each.get("channel_name"),
            "art": {
                "thumb": IMG_CATCHUP + each.get("logoUrl"),
                "icon": IMG_CATCHUP + each.get("logoUrl"),
                "fanart": IMG_CATCHUP + each.get("logoUrl"),
                "clearlogo": IMG_CATCHUP + each.get("logoUrl"),
                "clearart": IMG_CATCHUP + each.get("logoUrl"),
            },
            "callback": play,
            "params": {
                "channel_id": each.get("channel_id")
            }
        })
        if each.get("isCatchupAvailable"):
            litm.context.container(show_epg, "Catchup",
                                   0, each.get("channel_id"))
        yield litm


# Shows EPG container from Context menu
@Route.register
def show_epg(plugin, day, channel_id):
    resp = urlquick.get(CATCHUP_SRC.format(day, channel_id), max_age=-1).json()
    epg = sorted(
        resp['epg'], key=lambda show: show['startEpoch'], reverse=True)
    livetext = '[COLOR red] [ LIVE ] [/COLOR]'
    for each in epg:
        current_epoch = int(time()*1000)
        if not each['stbCatchupAvailable'] or each['startEpoch'] > current_epoch:
            continue
        islive = each['startEpoch'] < current_epoch and each['endEpoch'] > current_epoch
        showtime = '   '+livetext if islive else datetime.fromtimestamp(
            int(each['startEpoch']*.001)).strftime('    [ %I:%M %p -') + datetime.fromtimestamp(int(each['endEpoch']*.001)).strftime(' %I:%M %p ]   %a')
        yield Listitem.from_dict(**{
            "label": each['showname'] + showtime,
            "art": {
                'thumb': IMG_CATCHUP_SHOWS+each['episodePoster'],
                'icon': IMG_CATCHUP_SHOWS+each['episodePoster'],
                'fanart': IMG_CATCHUP_SHOWS+each['episodePoster'],
            },
            "callback": play,
            "info": {
                'title': each['showname'] + showtime,
                'originaltitle': each['showname'],
                "tvshowtitle": each['showname'],
                'genre': each['showGenre'],
                'plot': each['description'],
                "episodeguide": each.get("episode_desc"),
                'episode': 0 if each['episode_num'] == -1 else each['episode_num'],
                'cast': each['starCast'].split(', '),
                'director': each['director'],
                'duration': each['duration']*60,
                'tag': each['keywords'],
                'mediatype': 'episode',
            },
            "params": {
                "channel_id": each.get("channel_id"),
                "showtime": None if islive else each.get("showtime", "").replace(":", ""),
                "srno": None if islive else datetime.fromtimestamp(int(each.get("startEpoch", 0)*.001)).strftime('%Y%m%d')
            }
        })
    if int(day) == 0:
        for i in range(-1, -7, -1):
            label = 'Yesterday' if i == - \
                1 else (date.today() + timedelta(days=i)).strftime('%A %d %B')
            yield Listitem.from_dict(**{
                "label": label,
                "callback": Route.ref("/resources/lib/main:show_epg"),
                "params": {
                    "day": i,
                    "channel_id": channel_id
                }
            })


@Resolver.register
@isLoggedIn
def play_ex(plugin, dt=None):
    is_helper = inputstreamhelper.Helper(
        dt.get("proto", "mpd"), drm=dt.get("drm"))
    if is_helper.check_inputstream():
        licenseUrl = dt.get("lUrl") and dt.get("lUrl").replace("{HEADERS}", urlencode(
            getHeaders())).replace("{TOKEN}", urlencode(getTokenParams()))
        art = {}
        if dt.get("default_logo"):
            art['thumb'] = art['icon'] = IMG_CATCHUP + \
                dt.get("default_logo")
        return Listitem().from_dict(**{
            "label": dt.get("label") or plugin._title,
            "art": art or None,
            "callback": dt.get("pUrl"),
            "properties": {
                "IsPlayable": True,
                "inputstream": is_helper.inputstream_addon,
                "inputstream.adaptive.stream_headers": dt.get("hdrs"),
                "inputstream.adaptive.manifest_type": dt.get("proto", "mpd"),
                "inputstream.adaptive.license_type": dt.get("drm"),
                "inputstream.adaptive.license_key": licenseUrl,
            }
        })


# Play live stream/ catchup according to params.
# Also insures that user is logged in.
@Resolver.register
@isLoggedIn
def play(plugin, channel_id, showtime=None, srno=None):
    is_helper = inputstreamhelper.Helper("mpd", drm="com.widevine.alpha")
    hasIs = is_helper.check_inputstream()
    if not hasIs:
        return
    if showtime is None and Settings.get_boolean("extra"):
        with open(EXTRA_CHANNELS, "r") as f:
            extra = json.load(f)
        if extra.get(str(channel_id)):
            if extra.get(str(channel_id)).get("ext"):
                return extra.get(str(channel_id)).get("ext")
            return PLAY_EX_URL + extra.get(str(channel_id)).get("data")
    db = getHeaders()
    srno = datetime.now().strftime('%y%m%d%H%M%S')
    # Script.log("###################################### PLAY CHANNEL #######################################", lvl=Script.ERROR)
    headers = {
        "accesstoken":  db.get("authToken",""),
        "appkey":db.get("appkey",""),
        "camid": "",
        "channel_id":str(channel_id),
        "content-type":"application/x-www-form-urlencoded",
        "crmid":db.get("crmid",""),
        "deviceid":db.get("deviceId",""),
        "devicetype":"phone",
        "dm":"OnePlus ONEPLUS A5000",
        "isott":"false",
        "langid":"",
        "languageid":"6",
        "lbcookie":"1",
        "os":"android",
        "osversion":"10",
        "sid":db.get("uniqueId",""),
        "subscriberid":db.get("crmid",""),
        "uniqueid":db.get("uniqueId",""),
        "user-agent":"okhttp/4.2.2",
        "usergroup":db.get("usergroup",""),
        "userid":db.get("userId",""),
        "versioncode":"285",
    }

    body = {
        "channel_id": str(channel_id),
        "stream_type": "Seek"
    }
    if showtime and srno:
        body["begin"] = showtime
        body["srno"] = srno

    # Script.log(GET_CHANNEL_URL, lvl=Script.ERROR)
    # Script.log(headers, lvl=Script.ERROR)
    # Script.log(body, lvl=Script.ERROR)
    refresh_token()
    with no_ssl_verification():
        resp = urlquick.post(GET_CHANNEL_URL, data=body, headers=headers, max_age=0, verify=False, raise_for_status=False)
        if resp.status_code == 419 or resp.status_code == 403:
            # Login Again or Refresh
            refresh_token()
            executebuiltin("RunPlugin(plugin://plugin.video.jiotv/resources/lib/main/play/?channel_id={0})".format(str(channel_id)))
            return

    resp = resp.json()        
    # Script.log(resp, lvl=Script.ERROR)

    art = {}
    art["thumb"] = art["icon"] = IMG_CATCHUP + \
        resp.get("result", "").split("/")[-1].replace(".m3u8", ".png")
    params = getTokenParams()
    playback_url = resp.get("result","")

    playback_headers = {
        "accept-encoding":"gzip, deflate",
        "accesstoken": db.get("authToken",""),
        "channelid":str(channel_id),
        "crmid": db.get("crmid",""),
        "deviceid":db.get("deviceId",""),
        "devicetype":"phone",
        "os":"android",
        "osversion":"10",
        "srno": srno,
        "ssotoken": db.get("ssotoken",""),
        "subscriberid":db.get("crmid",""),
        "uniqueid":db.get("uniqueId",""),
        "user-agent":"plaYtv/7.0.8 (Linux;Android 10) ExoPlayerLib/2.11.7",
        "usergroup":db.get("usergroup",""),
        "userid":db.get("userId",""),
        "versioncode":"285",
    }

    #master playlist cookies
    mresp = requests.get(playback_url, headers=playback_headers, max_age=0, verify=False)
    playback_headers["Cookie"] = mresp.headers.get("set-cookie","")

    variant_m3u8 = m3u8.load(playback_url,headers=playback_headers)
    if variant_m3u8.is_variant:
        playback_url = variant_m3u8.playlists[-1].absolute_uri

    return Listitem().from_dict(**{
        "label": plugin._title,
        "art": art,
        "callback": playback_url,
        "properties": {
            "IsPlayable": True,
            "inputstream": "inputstream.adaptive",
            "inputstream.adaptive.stream_headers": urlencode(playback_headers),
            "inputstream.adaptive.manifest_type": "hls",
            "inputstream.adaptive.license_key": "|" + urlencode(playback_headers) + "|R{SSM}|",
        }
    })


# Login `route` to access from Settings
@Script.register
def login(plugin):
    headers = getHeaders()
    if headers:
        refresh_token()
        return
    mobile = keyboard("Enter your Jio mobile number")
    error = sendOTP(mobile)
    if error:
        Script.notify("Login Error", error)
        return
    otp = keyboard("Enter OTP", hidden=True)
    ULogin(mobile, otp, mode="otp")


# Logout `route` to access from Settings
@Script.register
def logout(plugin):
    ULogout()


# M3u Generate `route`
@Script.register
@isLoggedIn
def m3ugen(plugin, notify="yes"):
    refresh_token()
    channels = urlquick.get(CHANNELS_SRC).json().get("result")
    m3ustr = "#EXTM3U x-tvg-url=\"%s\"" % EPG_SRC
    for i, channel in enumerate(channels):
        lang = LANG_MAP[channel.get("channelLanguageId")]
        genre = GENRE_MAP[channel.get("channelCategoryId")]
        if not Settings.get_boolean(lang):
            continue
        group = lang + ";" + genre
        _play_url = PLAY_URL + \
            "channel_id={0}".format(channel.get("channel_id"))
        catchup = ""
        if channel.get("isCatchupAvailable"):
            catchup = ' catchup="vod" catchup-source="{0}channel_id={1}&showtime={{H}}{{M}}{{S}}&srno={{Y}}{{m}}{{d}}" catchup-days="7"'.format(
                PLAY_URL, channel.get("channel_id"))
        m3ustr += M3U_CHANNEL.format(
            tvg_id=channel.get("channel_id"),
            channel_name=channel.get("channel_name"),
            group_title=group,
            tvg_chno=int(channel.get("channel_order", i))+1,
            tvg_logo=IMG_CATCHUP + channel.get("logoUrl", ""),
            catchup=catchup,
            play_url=_play_url,
        )
    with open(M3U_SRC, "w+") as f:
        f.write(m3ustr.replace(u'\xa0', ' ').encode('utf-8').decode('utf-8'))
    if notify == "yes":
        Script.notify(
            "JioTV", "Playlist updated. Restart to apply the changes.")


# PVR Setup `route` to access from Settings
@Script.register
def pvrsetup(plugin):
    executebuiltin(
        "RunPlugin(plugin://plugin.video.jiotv/resources/lib/main/m3ugen/)")
    IDdoADDON = 'pvr.iptvsimple'

    def set_setting(id, value):
        if Addon(IDdoADDON).getSetting(id) != value:
            Addon(IDdoADDON).setSetting(id, value)
    if check_addon(IDdoADDON):
        set_setting("m3uPathType", "0")
        set_setting("m3uPath", M3U_SRC)
        set_setting("epgPathType", "1")
        set_setting("epgUrl", EPG_SRC)
        set_setting("catchupEnabled", "true")
        set_setting("catchupWatchEpgBeginBufferMins", "0")
        set_setting("catchupWatchEpgEndBufferMins", "0")


# Cache cleanup
@Script.register
def cleanup(plugin):
    urlquick.cache_cleanup(-1)
    Script.notify("Cache Cleaned", "")
